### ISSUE TITLE
build: Failed to build libass: command execution failed on 10.5.8 ppc

### ISSUE THREAD

**Issue by saxfun** [libass/libass#826]
### libass version

0.17.2

### How did you invoke configure?

sudo port install libass

### System Information

Darwin  9.8.0 Darwin Kernel Version 9.8.0: Wed Jul 15 16:57:01 PDT 2009; root:xnu-1228.15.4~1/RELEASE_PPC Power Macintosh 
Leopard 10.5.8, Macports 2.10.1, libass 0.17.2

### Toolchain

/opt/local/bin/gcc-mp-7 -v

/opt/local/bin/gcc-mp-7 -v
Using built-in specs.
COLLECT_GCC=/opt/local/bin/gcc-mp-7
COLLECT_LTO_WRAPPER=/opt/local/libexec/gcc/ppc-apple-darwin9/7.5.0/lto-wrapper
Target: ppc-apple-darwin9
Configured with: /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_gcc7/gcc7/work/gcc-7.5.0/configure --prefix=/opt/local --build=ppc-apple-darwin9 --enable-languages=c,c++,objc,obj-c++,lto,fortran --libdir=/opt/local/lib/gcc7 --includedir=/opt/local/include/gcc7 --infodir=/opt/local/share/info --mandir=/opt/local/share/man --datarootdir=/opt/local/share/gcc-7 --with-local-prefix=/opt/local --with-system-zlib --disable-nls --program-suffix=-mp-7 --with-gxx-include-dir=/opt/local/include/gcc7/c++/ --with-gmp=/opt/local --with-mpfr=/opt/local --with-mpc=/opt/local --with-isl=/opt/local --enable-stage1-checking --disable-multilib --enable-lto --enable-libstdcxx-time --with-build-config=bootstrap-debug --with-as=/opt/local/bin/as --with-ld=/opt/local/bin/ld --with-ar=/opt/local/bin/ar --with-bugurl=https://trac.macports.org/newticket --disable-tls --with-pkgversion='MacPorts gcc7 7.5.0_4'
Thread model: posix
gcc version 7.5.0 (MacPorts gcc7 7.5.0_4) 

ld -v
@(#)PROGRAM:ld  PROJECT:ld64-97.17
configured to support archs: i386 x86_64 ppc ppc64 armv6 armv7
LTO support using: LLVM version 3.3

### Log of the error

[main.txt](https://github.com/user-attachments/files/16991588/main.txt)

### Configuration log

:info:build libtool: compile:  /opt/local/bin/gcc-mp-7 -DHAVE_CONFIG_H -I. -I./libass -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT libass/ass_coretext.lo -MD -MP -MF libass/.deps/ass_coretext.Tpo -c libass/ass_coretext.c  -fno-common -DPIC -o libass/.libs/ass_coretext.o
:info:build libass/ass_coretext.c:37:18: error: missing binary operator before token "("
:info:build  #if __has_builtin(__builtin_available)
:info:build                   ^
:info:build libass/ass_coretext.c: In function 'get_font_file':
:info:build libass/ass_coretext.c:96:30: error: 'kCTFontURLAttribute' undeclared (first use in this function); did you mean 'kCTFontSizeAttribute'?
:info:build      else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
:info:build                               ^
:info:build libass/ass_coretext.c:42:50: note: in definition of macro 'CHECK_AVAILABLE'
:info:build  #define CHECK_AVAILABLE(sym, ...) (!!(intptr_t) &sym)
:info:build                                                   ^~~
:info:build libass/ass_coretext.c:96:30: note: each undeclared identifier is reported only once for each function it appears in
:info:build      else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
:info:build                               ^
:info:build libass/ass_coretext.c:42:50: note: in definition of macro 'CHECK_AVAILABLE'
:info:build  #define CHECK_AVAILABLE(sym, ...) (!!(intptr_t) &sym)
:info:build                                                   ^~~
:info:build make[1]: *** [libass/ass_coretext.lo] Error 1

**Comment by saxfun** [comment]
problems here with libass, macports 2.10 on PPC 10.5.8.
any hints on fixing libasss for compiling ?

thanks, saxfun :)

**Comment by barracuda156** [comment]
@saxfun Please edit the title to reflect that the issue happens on 10.5.8.

**Comment by barracuda156** [comment]
Also, the actual error is:
```
:info:build libtool: compile:  /opt/local/bin/gcc-mp-7 -DHAVE_CONFIG_H -I. -I./libass -I/opt/local/include -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter -Werror-implicit-function-declaration -Wstrict-prototypes -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers -D_GNU_SOURCE -pipe -Os -arch ppc -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/fribidi -I/opt/local/include/harfbuzz -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -I/opt/local/include/freetype2 -I/opt/local/include/libpng16 -MT libass/ass_coretext.lo -MD -MP -MF libass/.deps/ass_coretext.Tpo -c libass/ass_coretext.c  -fno-common -DPIC -o libass/.libs/ass_coretext.o
:info:build libass/ass_coretext.c:37:18: error: missing binary operator before token "("
:info:build  #if __has_builtin(__builtin_available)
:info:build                   ^
:info:build libass/ass_coretext.c: In function 'get_font_file':
:info:build libass/ass_coretext.c:96:30: error: 'kCTFontURLAttribute' undeclared (first use in this function); did you mean 'kCTFontSizeAttribute'?
:info:build      else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
:info:build                               ^
:info:build libass/ass_coretext.c:42:50: note: in definition of macro 'CHECK_AVAILABLE'
:info:build  #define CHECK_AVAILABLE(sym, ...) (!!(intptr_t) &sym)
:info:build                                                   ^~~
:info:build libass/ass_coretext.c:96:30: note: each undeclared identifier is reported only once for each function it appears in
:info:build      else if (CHECK_AVAILABLE(kCTFontURLAttribute, macOS 10.6, *)) {
:info:build                               ^
:info:build libass/ass_coretext.c:42:50: note: in definition of macro 'CHECK_AVAILABLE'
:info:build  #define CHECK_AVAILABLE(sym, ...) (!!(intptr_t) &sym)
:info:build                                                   ^~~
:info:build make[1]: *** [libass/ass_coretext.lo] Error 1
```

**Comment by barracuda156** [comment]
`kCTFontURLAttribute` is indeed present in 10.6+ (apparently, already the early developer builds of 10.6 have it).
https://developer.apple.com/documentation/coretext/kctfonturlattribute
It should not be used on 10.4–10.5.

**Comment by barracuda156** [comment]
@astiob Could you please help here a bit? I am not really sure why this fails. Looking at the commit https://github.com/libass/libass/commit/5c15c883a4783641f7e71a6a1f440209965eb64f it seems that on 10.5 that chunk should be skipped, but it is not :(

**Comment by TheOneric** [comment]
Can you replace
```C
#ifndef __has_builtin
#define __has_builtin 0
#endif
```

near the start of `libass/ass_coretext.c` with

```C
#ifndef __has_builtin
#define __has_builtin(x) 0
#endif
```

and attempt to build again?

**Comment by astiob** [comment]
Hm, didn’t we have that `#define` fix in some PR? Is it one of those stuck in review hell? Or did we have multiple copies and only fixed one?

**Comment by barracuda156** [comment]
@astiob I cannot check on 10.5 now, but on 10.6 ppc with gcc14 I can build 0.17.2 without any errors from the default portfile. However, @saxfun uses gcc7, which might have not been tested earlier.

**Comment by TheOneric** [comment]
We recently gained a second fallback copy for `__has_builtin` in `ass_utils.h` together with the `ASSUME` macro; the new one already uses `(x)`. Note this doesn't exist in the 0.17.2 *(neither in 0.17.3)* tarball used here.
If there was an attempt to fix this before, I don't remember it anymore.

**Comment by astiob** [comment]
> Or did we have multiple copies and only fixed one?

Ah, apparently #820 _added_ a second, correct copy but didn’t remove the one in `ass_coretext`. But it should be shadowed by the added one, so master should compile fine. But OP is trying to compile 0.17.2, which doesn’t include #820 yet.

**Comment by TheOneric** [comment]
#820’s copy is in `ass_utils.h` and neither `ass_coretext.c` nor `ass_coretext.h` directly include `ass_utils.h` afaict. If it get’s shadowed then only by coincidence.
Perhaps we should move this to `ass_compat.h`?

**Comment by saxfun** [comment]
Thanks a lot. Due to your new commits libass is compiling fine on ppc 10.5.8 again! super!

**Comment by barracuda156** [comment]
@astiob @TheOneric Thank you for fixing this!

**Comment by astiob** [comment]
> By the way, let me update it in MacPorts to 0.17.3, but hopefully with an added fix for 10.5 + gcc7.

FYI: the simple change Oneric suggested [above](https://github.com/libass/libass/issues/826#issuecomment-2349299647) should be enough to fix 0.17.2 or 0.17.3.

**Comment by barracuda156** [comment]
> > By the way, let me update it in MacPorts to 0.17.3, but hopefully with an added fix for 10.5 + gcc7.
> 
> FYI: the simple change Oneric suggested [above](https://github.com/libass/libass/issues/826#issuecomment-2349299647) should be enough to fix 0.17.2 or 0.17.3.

@astiob Thank you. I borrowed two commits, and `libass-devel` updated to master including those: https://github.com/macports/macports-ports/commit/0e6b5de75b9b5c40957c7c9fa8794a8f7208f98a

Hopefully that works on 10.5 (I cannot test), but on 10.6 ppc all builds fine.
